### PR TITLE
WIP: Theano variable arguments for pooling

### LINKED ISSUE #3349: downsample.max_pool_2d only accepts int argument
Currently downsample.max_pool_2d only accepts int argument for ds and st parameters. However it will be better if it can also accept tensor variable, for example in spatial pyramid pooling, the st and ds changes when the input image size changes.

```python
        b, c, h, w = X.shape  # tensor variable
        layer_out = []
        for i in [1,2,4]:
            out = downsample.max_pool_2d(state_below, ds=(h/i, w/i),
                                         st=(h/i, w/i),
                                         ignore_border=self.ignore_border,
                                         mode=self.mode)

            out = out.reshape((b, T.prod(out.shape)/b))
            layer_out.append(out)
```
here we get the error that max_pool_2d only takes integer value for ds and st. I think it will be better for ds and st to accept both integer and tensor variable so the max_pool_2d can change on the fly in the computation graph, if theano thinks this is useful, i can work on a pull request

### PR DESCRIPTION
**Pull request by Thrandis**
Fixes  #3349

**TODO**:
- [x] Change the interface of the op and the grads.
- [x] Adapt the "checking" of the arguments (assert that they are integers, ...).
- [x] Add the `prepare_node` for backward compatibility.
- [x] Write the tests when passing shared variables.
- [x] Write the test for the old interface.
- [x] Update the docstrings.
- [x] Change the c code.
- [x] Make all tests pass.
- [x] Delete `downsample.py`
- [ ] Adapt Theano to the new `Pool` interface
- [x] Merge the code.

**Questions**:

1. @nouiz @mgermain I have 3 errors left in the original tests, all caused by `DownsampleFactorMaxGrad` (from [downsample.py](https://github.com/Theano/Theano/blob/master/theano/tensor/signal/downsample.py)) Shall I also change its interface?
2. Now that we can pass the window size, padding size, etc. as tensor variable, what kind of checking shall we do on them (like are they of the integer type?)?

### CODE REVIEW

**Comment by nouiz** [comment]
Theano automatically check if the input tensor have the right dtype and ndim of dimensions.

I think you also should change the grad op. I didn't read the code.

**Comment by Thrandis** [comment]
@nouiz alright! About the grad, it is just that the current interface is like this because:
""""
# This is for compatibility with pickled things.  It should go away at
# some point.
""""
So if I change it, it will probably remove some backward compatibility features.

File: theano/tensor/signal/pool.py [hunk 1]
```diff
@@ -235,49 +234,54 @@ def out_shape(imgshape, ds, ignore_border=False, st=None, padding=(0, 0)):
...
             raise ValueError(
                 "Pool mode parameter only support 'max', 'sum',"
                 " 'average_inc_pad' and 'average_exc_pad'. Got %s" % mode)
         self.mode = mode
 
-    def make_node(self, x):
+    def make_node(self, x, ws, stride=None, pad=(0, 0)):
         # TODO: consider restricting the dtype?
         x = tensor.as_tensor_variable(x)
+        # TODO CESAR: How can we check the theano variables?
```
Comment (mgermain) [inline theano/tensor/signal/pool.py]: For the type of theano var it's 'ws.type'.

**Comment by mgermain** [comment]
Have you run the test with the new backend as well ?

**Comment by Thrandis** [comment]
> Have you run the test with the new backend as well ?

No :-)

**Comment by mgermain** [comment]
This old PR is related to what you are are going right now.
https://github.com/Theano/Theano/pull/3965

You can use `prepare_node` to avoid breaking backward compatibility with the pickling.
@nouiz should confirm

**Comment by Thrandis** [comment]
Here is the code that I used to craete `old_pool_interface.pkl` (so that we can retrieve it if needed):
```python
import theano

from theano.tensor.signal.pool import pool_2d
from six.moves import cPickle

x = theano.tensor.ftensor4()
y = pool_2d(x, (2, 2), mode='max', ignore_border=True)
z = pool_2d(x, (2, 2), mode='average_exc_pad', ignore_border=True)

dy_dx = theano.gradient.grad(y.sum(), x)
dz_dx = theano.gradient.grad(z.sum(), x)

fct = theano.function([x], [y, z, dy_dx, dz_dx])

with open('old_pool_interface.pkl', 'wb') as f:
    cPickle.dump(fct, f, protocol=cPickle.HIGHEST_PROTOCOL)

```

**Comment by Thrandis** [comment]
# Errors left (with python code)

/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")
.....E.....E......ERROR (theano.gof.opt): EquilibriumOptimizer max'ed out by 'local_upcast_elemwise_constant_inputs'. You can safely raise the current threshold of 5.000000 with the theano flag 'optdb.max_use_ratio'.
ERROR (theano.gof.opt): EquilibriumOptimizer max'ed out by 'local_upcast_elemwise_constant_inputs'. You can safely raise the current threshold of 5.000000 with the theano flag 'optdb.max_use_ratio'.
ERROR (theano.gof.opt): EquilibriumOptimizer max'ed out by 'local_upcast_elemwise_constant_inputs'. You can safely raise the current threshold of 5.000000 with the theano flag 'optdb.max_use_ratio'.
ERROR (theano.gof.opt): EquilibriumOptimizer max'ed out by 'local_upcast_elemwise_constant_inputs'. You can safely raise the current threshold of 5.000000 with the theano flag 'optdb.max_use_ratio'.
ERROR (theano.gof.opt): EquilibriumOptimizer max'ed out by 'local_upcast_elemwise_constant_inputs'. You can safely raise the current threshold of 5.000000 with the theano flag 'optdb.max_use_ratio'.

======================================================================
ERROR: test_DownsampleFactorMaxGrad (__main__.TestDownsampleFactorMax)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_pool.py", line 841, in test_DownsampleFactorMaxGrad
    mode=mode)(im, maxout,
TypeError: __new__() takes at least 3 arguments (3 given)

======================================================================
ERROR: test_DownsampleFactorMaxPaddingStride_grad_grad (__main__.TestDownsampleFactorMax)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_pool.py", line 585, in test_DownsampleFactorMaxPaddingStride_grad_grad
    utt.verify_grad(mp, [imval, grad_val], rng=rng)
  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/tests/unittest_tools.py", line 91, in verify_grad
    T.verify_grad(op, pt, n_tests, rng, *args, **kwargs)
  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gradient.py", line 1685, in verify_grad
    analytic_grad = grad_fn(*[p.copy() for p in pt])
  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/compile/function_module.py", line 912, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gof/link.py", line 314, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/compile/function_module.py", line 899, in __call__
    self.fn() if output_subset is None else\
  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gof/op.py", line 905, in rval
    r = p(n, [x[0] for x in i], o)
  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/tensor/signal/pool.py", line 1063, in perform
    if (maxout[n, k, r, c] == y_padded[n, k, row_ind, col_ind]):
IndexError: ('index 10 is out of bounds for axis 3 with size 10\nApply node that caused the error: DownsampleFactorMaxGradGrad{ignore_border=True, mode=\'max\'}(input 0, Pool{ignore_border=True, mode=\'max\'}.0, InplaceDimShuffle{x,x,0,1}.0, TensorConstant{[5 3]}, TensorConstant{[3 2]}, TensorConstant{(2,) of 2})\nToposort index: 7\nInputs types: [TensorType(float64, (True, True, False, False)), TensorType(float64, (True, True, False, False)), TensorType(float64, (True, True, False, False)), TensorType(int64, vector), TensorType(int64, vector), TensorType(int64, vector)]\nInputs shapes: [(1, 1, 10, 10), (1, 1, 4, 6), (1, 1, 10, 10), (2,), (2,), (2,)]\nInputs strides: [(800, 800, 80, 8), (192, 192, 48, 8), (800, 800, 80, 8), (8,), (8,), (8,)]\nInputs values: [\'not shown\', \'not shown\', \'not shown\', array([5, 3]), array([3, 2]), array([2, 2])]\nOutputs clients: [[\'output\']]\n\nBacktrace when the node is created(use Theano flag traceback.limit=N to make it longer):\n  File "/Tmp/lisa/os_v5/anaconda/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n  File "test_pool.py", line 585, in test_DownsampleFactorMaxPaddingStride_grad_grad\n    utt.verify_grad(mp, [imval, grad_val], rng=rng)\n  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/tests/unittest_tools.py", line 91, in verify_grad\n    T.verify_grad(op, pt, n_tests, rng, *args, **kwargs)\n  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gradient.py", line 1675, in verify_grad\n    disconnected_inputs=\'ignore\')\n  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gradient.py", line 549, in grad\n    grad_dict, wrt, cost_name)\n  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gradient.py", line 1312, in _populate_grad_dict\n    rval = [access_grad_cache(elem) for elem in wrt]\n  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gradient.py", line 1267, in access_grad_cache\n    term = access_term_cache(node)[idx]\n  File "/u/laurent/.local/lib/python2.7/site-packages/Theano-0.9.0.dev1-py2.7.egg/theano/gradient.py", line 1101, in access_term_cache\n    input_grads = node.op.grad(inputs, new_output_grads)\n\nHINT: Use the Theano flag \'exception_verbosity=high\' for a debugprint and storage map footprint of this apply node.', '\nThe error happened with the following inputs:', [array([[[[ 7.00437122,  8.44186643,  6.76514336,  7.27858057,  9.51457957,
           0.12703197,  4.13587699,  0.48812794,  0.99928561,  5.08066306],
         [ 2.00247539,  7.44154169,  1.92892003,  7.00844752,  2.93228106,
           7.74479454,  0.05108839,  1.12857654,  1.10953672,  2.47668229],
         [ 0.23236299,  7.27321154,  3.40034942,  1.97503156,  9.09179593,
           9.78346985,  5.32802544,  2.59131849,  5.83812619,  3.25690653],
         [ 8.88899314,  6.26404527,  8.18873691,  5.4734542 ,  4.16712006,
           7.43047192,  3.69596381,  0.75166544,  7.75192978,  2.19409244],
         [ 0.79342127,  4.86780522,  1.53673902,  8.28465133,  1.91368568,
           2.70408954,  5.61034421,  9.0238039 ,  8.51788342,  4.18081961],
         [ 3.93476269,  0.16220513,  2.99213372,  3.53778219,  8.93502665,
           7.86136569,  7.71386926,  4.20054857,  7.76025138,  4.64308144],
         [ 1.81770166,  8.84025605,  7.18792268,  6.718813  ,  2.56563633,
           4.30801822,  0.16453576,  2.34993827,  5.11171311,  2.9200924 ],
         [ 5.01893508,  4.9827313 ,  1.03771524,  4.46443115,  9.69189168,
           7.38471122,  7.19550608,  8.93043387,  9.62674681,  1.97050228],
         [ 7.14589964,  1.61923939,  8.66254772,  6.23820247,  9.59455124,
           5.24142042,  0.36432878,  7.26871579,  0.03909844,  0.50294001],
         [ 9.91992316,  2.122575  ,  9.47370665,  4.51540549,  9.98794673,
           6.47501489,  7.0224071 ,  4.29581765,  1.6777476 ,  1.18133093]]]]), array([[[[ 9.37882622,  9.10997444,  3.97993537,  9.25303266,  7.45499217,
           7.89826364],
         [ 6.37708518,  9.92741367,  4.76105537,  1.02114186,  8.72563325,
           7.56127044],
         [ 2.99198836,  1.86256519,  4.94353514,  0.40895499,  4.78352282,
           3.08990199],
         [ 9.20580262,  9.17945685,  4.68488931,  2.34313083,  0.89824944,
           6.79403574]]]])], '\nThe value of eps is:', None, '\nThe out_type is:', None)

----------------------------------------------------------------------
Ran 25 tests in 230.549s

FAILED (errors=2)


So the first one is about `DownsampleFactorMaxGrad` in `donwsample.py`. It is an old interface, kept here for backward compatibility reasons. So just to confirm: Can I remove it?

The second error I'm not sure what it is happening. Also, what is this error:  `ERROR (theano.gof.opt): EquilibriumOptimizer max'ed out`?

**Comment by Thrandis** [comment]
@lamblin

**Comment by Thrandis** [comment]
I corrected the c code. The same 2 tests are failing, plus the test of the old interface. I'll look into it.

**Comment by Thrandis** [comment]
I fixed the `test_DownsampleFactorMaxPaddingStride_grad_grad`.

File: theano/tensor/signal/pool.py [hunk 2]
```diff
@@ -944,14 +1039,14 @@ def perform(self, node, inp, out):
...
-        pd0, pd1 = self.padding
+        ws0, ws1 = ws
+        st0, st1 = stride
+        pd0, pd1 = pad
         img_rows = x.shape[-2] + 2 * pd0
         img_cols = x.shape[-1] + 2 * pd1
 
         # pad the image and its gradients
-        if self.padding != (0, 0):
+        if pd0 != 0 and pd1 != 0:
```
Comment (lamblin) [inline theano/tensor/signal/pool.py]: Should it be `pd0 != 0 or pd1 != 0` instead?

File: theano/tensor/signal/pool.py [hunk 3]
```diff
@@ -567,8 +571,8 @@ def c_code(self, node, name, inp, out, sub):
         return ccode % locals()
 
     def c_code_cache_version(self):
-        return (0, 6, 8, 4)
-
+        #return (0, 6, 8, 4)
+        return (0, 6, 8, 6)
```
Comment (lamblin) [inline theano/tensor/signal/pool.py]: OK, you can remove the commented-out version

File: theano/tensor/signal/pool.py [hunk 4]
```diff
@@ -878,8 +887,8 @@ def c_code(self, node, name, inp, out, sub):
         """ % locals()
 
     def c_code_cache_version(self):
-        return (0, 7)
-
+        #return (0, 7)
+        return (0, 9)
```
Comment (lamblin) [inline theano/tensor/signal/pool.py]: Same here.

File: theano/tensor/signal/pool.py [hunk 5]
```diff
@@ -1107,23 +1121,23 @@ def c_code(self, node, name, inp, out, sub):
...
               for(int k=0; k<PyArray_DIMS(%(x)s)[1]; k++){
                 for(int i=0; i< z_r; i++){
-                  r_st = i * %(st0)s;
-                  r_end = r_st + %(ws0)s;
-                  // skip the padding
-                  r_st = r_st < %(pd0)s ? %(pd0)s : r_st;
-                  r_end = r_end > (r - %(pd0)s) ? r - %(pd0)s : r_end;
+                  r_st = i * st0;
+                  r_end = r_st + ws0;
+                  // skip the paddin_g
```
Comment (lamblin) [inline theano/tensor/signal/pool.py]: ?

File: theano/tensor/signal/pool.py [hunk 6]
```diff
@@ -1148,4 +1162,5 @@ def c_code(self, node, name, inp, out, sub):
         """ % locals()
 
     def c_code_cache_version(self):
-        return (0, 1)
+        #return (0, 1)
+        return (0, 3)
```
Comment (lamblin) [inline theano/tensor/signal/pool.py]: Idem here.

File: theano/tensor/signal/pool.py (hunk 2 above)
Comment (Thrandis) [inline theano/tensor/signal/pool.py]: Both of them have to be non-zero to enter the if statement, right?

**Comment by Thrandis** [comment]
@lamblin So I fixed the test of the old interface and addressed your comments. Now what is left to do is:
- Remove `downsample.py`.
- Check if we need to change some parts of Theano to use the new Pool interface.
- Merge the code.
I'll do that when I come back from vacations.

**Comment by Thrandis** [comment]
Number of days since last merge fail: 0

**Comment by Thrandis** [comment]
Alright since the merging failed, I opened a new PR #4756

**Pull request closed without merging**